File: frame/map-frame-project/MapFrameCoreModule/src/main/java/com/grandtech/mapframe/core/net/BoxHttpClient.java
package com.grandtech.mapframe.core.net;

import android.util.Log;

import com.grandtech.mapframe.core.util.SimpleMap;
import com.mapbox.mapboxsdk.module.http.HttpRequestUtil;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import okhttp3.Dispatcher;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * @ClassName BoxHttpClient
 * @Description TODO 地图HttpClient
 * @Author: fs
 * @Date: 2021/5/11 13:16
 * @Version 2.0
 */
public final class BoxHttpClient {

    private OkHttpClient client;

    private SimpleMap<String, Long> expiresMap;

    private String mToken ;

    private static volatile BoxHttpClient singleton;

    private List<Interceptor> interceptors;

    private BoxHttpClient() {
        client = initBuilder().dispatcher(getDispatcher()).build();
        HttpRequestUtil.setOkHttpClient(client);
        expiresMap = new SimpleMap<>();
    }

    public static BoxHttpClient getSingleton() {
        if (singleton == null) {
            synchronized (BoxHttpClient.class) {
                if (singleton == null) {
                    singleton = new BoxHttpClient();
                }
            }
        }
        return singleton;
    }

    /**
     * 判断是否初始化
     * @return
     */
    public static boolean isInit() {
        if (singleton == null) {
           return false;
        }
        return true;
    }
    /**
     * 清空拦截器
     */
    public void clearInterceptor(){
        interceptors = null;
    }

    /**
     * 设置网络请求Token
     * @param mToken
     */
    public void setToken(String mToken) {
        this.mToken = mToken;
    }

    public BoxHttpClient addInterceptor(Interceptor interceptor) {
        if(interceptors == null){
            interceptors = new ArrayList<>();
        }
        interceptors.add(interceptor);
        client = initBuilder(interceptors).dispatcher(getDispatcher()).build();
        HttpRequestUtil.setOkHttpClient(client);
        return this;
    }

    public OkHttpClient getClient() {
        return client;
    }

    public void addLayerExpires(String layerName, Long expiresTime) {
        expiresMap.push(layerName, expiresTime);
    }

    public void removeLayerExpires(String layerName) {
        expiresMap.remove(layerName);
    }

    private OkHttpClient.Builder initBuilder() {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.addInterceptor(getInterceptor());
        return builder;
    }
    private OkHttpClient.Builder initBuilder(List<Interceptor> interceptors) {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.addInterceptor(getInterceptor());
        for (Interceptor interceptor : interceptors) {
            builder.addInterceptor(interceptor);
        }
        return builder;
    }

    private Dispatcher getDispatcher() {
        Dispatcher dispatcher = new Dispatcher();
        dispatcher.setMaxRequestsPerHost(15);
        return dispatcher;
    }

    public Interceptor getInterceptor() {
        return new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();
                String requestUrl = chain.request().url().toString();
                Log.i("gmbgl-BoxHttpClient",requestUrl);
                Long request_time = getStandardUtcTime();
                Long expires_time = dispatchExpiresTime(requestUrl);
                Request.Builder builder = request.newBuilder();
                builder.url(requestUrl);
                if(mToken!=null){
                    builder.addHeader("Authorization", mToken);
                }
                builder.addHeader("request_time", request_time + "");
                if (expires_time != null) {
                    builder.addHeader("expires_time", expires_time.toString());
                }

                Response response = chain.proceed(builder.build());
                return response;
            }
        };
    }
    private Long dispatchExpiresTime(String url) {
        if (expiresMap == null) {
            return null; //ms
        } else {
            for (String layerName : expiresMap.keySet()) {
                if (url.contains(layerName + "/")) {
                    return expiresMap.get(layerName);
                }
            }
        }
        return null; //ms
    }
    /**
     * utc时间
     *
     * @return
     */
    private  long getStandardUtcTime() {
        Calendar cal = Calendar.getInstance();
        long mills = cal.getTimeInMillis();
        return mills;
    }
}

File: frame/map-frame-project/MapFrameCoreModuleV10/src/main/java/com/gradtech/mapframev10/core/net/BoxHttpClient.kt
package com.gradtech.mapframev10.core.net

import android.util.Log
import com.mapbox.common.*
import java.util.*
import java.util.Collections.singleton

/**
 * @ClassName BoxHttpClient
 * @Description TODO 一些网络设置
 * @Author: fs
 * @Date: 2022/12/28 8:35
 * @Version 2.0
 */
class BoxHttpClient  : HttpServiceInterceptorInterface{


    companion object {
        private var instance: BoxHttpClient? = null
        @Synchronized
        fun instance(): BoxHttpClient {
            if (instance == null) instance = BoxHttpClient()
            return instance!!
        }
        fun cancel() {
            instance?.clearInterceptors();
            instance = null
        }
    }

    private var mInterceptors : ArrayList<HttpServiceInterceptorInterface> = arrayListOf();

    private val expiresMap:HashMap<String, Long> = hashMapOf<String, Long>()

    private var mToken : String? = null;

    private constructor(){
        HttpServiceFactory.getInstance().setInterceptor(this);
    }

    public fun addInterceptors(httpServiceInterceptorInterface: HttpServiceInterceptorInterface){
        mInterceptors.add(httpServiceInterceptorInterface);
    }

    fun clearInterceptors(){
        mInterceptors.clear();
    }

    fun addLayerExpires(layerName: String, expiresTime: Long) {
        expiresMap.put(layerName, expiresTime)
    }

    fun removeLayerExpires(layerName: String?) {
        expiresMap.remove(layerName)
    }

    /**
     * 设置网络请求Token
     * @param mToken
     */
    fun setToken(token: String) {
        this.mToken = token
    }

    override fun onRequest(request: HttpRequest): HttpRequest {
        val requestUrl: String = request.url;
        Log.i("gmbgl-BoxHttpClient", requestUrl)
        val request_time = getStandardUtcTime()
        val expires_time = dispatchExpiresTime(requestUrl)
        val headers = request.headers
        headers.put("request_time", request_time.toString() + "")
        if (expires_time != null) {
            headers.put("expires_time", expires_time.toString())
        }
        if (mToken != null) {
            headers.put("Authorization", mToken)
        }
        var build = request.toBuilder().headers(headers).build()
        //外部的拦截器
        for (mInterceptor in mInterceptors) {
            build = mInterceptor.onRequest(build)
        }
        return build
    }

    override fun onDownload(download: DownloadOptions): DownloadOptions {
        var download1: DownloadOptions = download;
        for (mInterceptor in mInterceptors) {
            download1 = mInterceptor.onDownload(download1)
        }
        return download1;
    }

    override fun onResponse(response: HttpResponse): HttpResponse {
        var response1: HttpResponse = response;
        for (mInterceptor in mInterceptors) {
            response1 = mInterceptor.onResponse(response1)
        }
        return response1
    }

    private fun dispatchExpiresTime(url: String): Long? {
        if (expiresMap == null) {
            return null //ms
        } else {
            for (layerName in expiresMap.keys) {
                if (url.contains("$layerName/")) {
                    return expiresMap.get(layerName)
                }
            }
        }
        return null //ms
    }

    /**
     * utc时间
     *
     * @return
     */
    private fun getStandardUtcTime(): Long {
        val cal = Calendar.getInstance()
        return cal.timeInMillis
    }
}